### PR TITLE
Adding logging

### PR DESCRIPTION
**Pull request by mcminis1**
#390

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@
 # For testing
 pytest==7.2.1
 pytest-dotenv==0.5.2
+hypothesis==6.67.1
```
Comment (jerryjliu) [inline requirements.txt:6]: why do we need this?

**Review by jerryjliu — COMMENTED** [review]
I think I'm ok with this! could you provide an example of a user being able to easily use the logger for stdout? I want to make sure that it's still easy to use in a Jupyter notebook setting.

also, are you planning to land multiple PR's?

File: requirements.txt (hunk 1 above)
Comment (mcminis1) [reply requirements.txt]: removed.

i thought it was a requirement for running the tests.

**Comment by mcminis1** [comment]
> I think I'm ok with this! could you provide an example of a user being able to easily use the logger for stdout? I want to make sure that it's still easy to use in a Jupyter notebook setting.

I'll try to add an example

> also, are you planning to land multiple PR's?

just one. this is currently a WIP draft.

**Comment by mcminis1** [comment]
@jerryjliu I updated a notebook, examples/vector_indices/SimpleIndexDemo.ipynb, with an example. Let me know what you think.

**Comment by jerryjliu** [comment]
thanks @mcminis1!

just a heads up that i do want to spend some time going through this PR to make sure that the user flows all make sense. i'll have some time to get to this later this week. apologies for the delay in advance, just wanted to let you know!

**Comment by mcminis1** [comment]
@jerryjliu sure. I like logging for production use cases. For the notebook, I agree, it feels a little weird. I think they two of us have slightly different perspectives on library design. I'm probably more concerned about use in an application and you're more interested in use in a notebook.

Usually objects have string methods `def __repr__(self) -> str:`  so that, in a notebook, if the object is the last thing in the cell it returns a string. The ntermediate progress isn't output, but the final thing is. If you want something as an intermediate output, then you would structure the object in such as way that you could get it. For instance, consider how requests has a [prepare method](https://requests.readthedocs.io/en/latest/user/advanced/#prepared-requests). Using that you can see all of the intermediate parts before executing the call.

So, after you think on it let me know what you want to do. Perhaps a good middle ground is leaving all the `verbose` stuff in the code and just adding in the logging as it is here. then the notebook experience is identical, but you can still get good logs.

**Comment by jerryjliu** [comment]
> @jerryjliu sure. I like logging for production use cases. For the notebook, I agree, it feels a little weird. I think they two of us have slightly different perspectives on library design. I'm probably more concerned about use in an application and you're more interested in use in a notebook.
> 
> Usually objects have string methods `def __repr__(self) -> str:` so that, in a notebook, if the object is the last thing in the cell it returns a string. The ntermediate progress isn't output, but the final thing is. If you want something as an intermediate output, then you would structure the object in such as way that you could get it. For instance, consider how requests has a [prepare method](https://requests.readthedocs.io/en/latest/user/advanced/#prepared-requests). Using that you can see all of the intermediate parts before executing the call.
> 
> So, after you think on it let me know what you want to do. Perhaps a good middle ground is leaving all the `verbose` stuff in the code and just adding in the logging as it is here. then the notebook experience is identical, but you can still get good logs.

Yeah I don't disagree. I do want to make this good for application use - mostly just trying to think about how to also keep the notebook experience. I will have some thoughts today or tomorrow!

**Comment by jerryjliu** [comment]
I just played around with it - actually I like this! I'm going to take a deeper look tonight. Thanks for putting together this PR @mcminis1, this is going to be super valuable

**Review by jerryjliu — COMMENTED** [review]
heads up: i'm pushing some changes to fix linting. afterwards i'm down to land! 

one followup after this, is that some logging uses the root logger whereas others uses the module logger. we should refactor to be consistent (and use module logger everywhere probably).

**Review by jerryjliu — APPROVED** [review]

**Comment by jerryjliu** [comment]
congrats! 

what's your twitter? happy to give you a shoutout on the release tomorrow

**Comment by mcminis1** [comment]
https://twitter.com/JeremyMcminis

thanks!

**Pull request merged**